### DISCUSSION TITLE
Unable to launch a duet

### DISCUSSION THREAD

**Discussion by raaghavvd** [OpenMined/courses#354] [Course: Foundations of Private Computation]
Hi, I'm trying to perform Duet basics and I'm getting the following error when I try to launch a duet 
---------------------------------------------------------------------------
TimeoutError                              Traceback (most recent call last)
/Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/site-packages/urllib3/connection.py in _new_conn(self)
    168         try:
--> 169             conn = connection.create_connection(
    170                 (self._dns_host, self.port), self.timeout, **extra_kw

/Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/site-packages/urllib3/util/connection.py in create_connection(address, timeout, source_address, socket_options)
     95     if err is not None:
---> 96         raise err
     97 

/Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/site-packages/urllib3/util/connection.py in create_connection(address, timeout, source_address, socket_options)
     85                 sock.bind(source_address)
---> 86             sock.connect(sa)
     87             return sock

TimeoutError: [Errno 60] Operation timed out

During handling of the above exception, another exception occurred:

NewConnectionError                        Traceback (most recent call last)
/Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/site-packages/urllib3/connectionpool.py in urlopen(self, method, url, body, headers, retries, redirect, assert_same_host, timeout, pool_timeout, release_conn, chunked, body_pos, **response_kw)
    698             # Make the request on the httplib connection object.
--> 699             httplib_response = self._make_request(
    700                 conn,

/Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/site-packages/urllib3/connectionpool.py in _make_request(self, conn, method, url, timeout, chunked, **httplib_request_kw)
    393             else:
--> 394                 conn.request(method, url, **httplib_request_kw)
    395 

/Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/site-packages/urllib3/connection.py in request(self, method, url, body, headers)
    233             headers["User-Agent"] = _get_default_user_agent()
--> 234         super(HTTPConnection, self).request(method, url, body=body, headers=headers)
    235 

/Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/http/client.py in request(self, method, url, body, headers, encode_chunked)
   1252         """Send a complete request to the server."""
-> 1253         self._send_request(method, url, body, headers, encode_chunked)
   1254 

/Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/http/client.py in _send_request(self, method, url, body, headers, encode_chunked)
   1298             body = _encode(body, 'body')
-> 1299         self.endheaders(body, encode_chunked=encode_chunked)
   1300 

/Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/http/client.py in endheaders(self, message_body, encode_chunked)
   1247             raise CannotSendHeader()
-> 1248         self._send_output(message_body, encode_chunked=encode_chunked)
   1249 

/Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/http/client.py in _send_output(self, message_body, encode_chunked)
   1007         del self._buffer[:]
-> 1008         self.send(msg)
   1009 

/Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/http/client.py in send(self, data)
    947             if self.auto_open:
--> 948                 self.connect()
    949             else:

/Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/site-packages/urllib3/connection.py in connect(self)
    199     def connect(self):
--> 200         conn = self._new_conn()
    201         self._prepare_conn(conn)

/Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/site-packages/urllib3/connection.py in _new_conn(self)
    180         except SocketError as e:
--> 181             raise NewConnectionError(
    182                 self, "Failed to establish a new connection: %s" % e

NewConnectionError: <urllib3.connection.HTTPConnection object at 0x7fa7c4a7e460>: Failed to establish a new connection: [Errno 60] Operation timed out

During handling of the above exception, another exception occurred:

MaxRetryError                             Traceback (most recent call last)
/Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/site-packages/requests/adapters.py in send(self, request, stream, timeout, verify, cert, proxies)
    438             if not chunked:
--> 439                 resp = conn.urlopen(
    440                     method=request.method,

/Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/site-packages/urllib3/connectionpool.py in urlopen(self, method, url, body, headers, retries, redirect, assert_same_host, timeout, pool_timeout, release_conn, chunked, body_pos, **response_kw)
    754 
--> 755             retries = retries.increment(
    756                 method, url, error=e, _pool=self, _stacktrace=sys.exc_info()[2]

/Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/site-packages/urllib3/util/retry.py in increment(self, method, url, response, error, _pool, _stacktrace)
    573         if new_retry.is_exhausted():
--> 574             raise MaxRetryError(_pool, url, error or ResponseError(cause))
    575 

MaxRetryError: HTTPConnectionPool(host='ec2-18-216-8-163.us-east-2.compute.amazonaws.com', port=5000): Max retries exceeded with url: //metadata (Caused by NewConnectionError('<urllib3.connection.HTTPConnection object at 0x7fa7c4a7e460>: Failed to establish a new connection: [Errno 60] Operation timed out'))

During handling of the above exception, another exception occurred:

ConnectionError                           Traceback (most recent call last)
<ipython-input-2-d5b9fe736ead> in <module>
----> 1 duet = sy.launch_duet(network_url="http://ec2-18-216-8-163.us-east-2.compute.amazonaws.com:5000/")

/Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/site-packages/syft-0.0.0-py3.9.egg/syft/grid/duet/__init__.py in launch_duet(logging, network_url, loopback, credential_exchanger, db_path)
    207     info("♫♫♫ > ...waiting for response from OpenGrid Network... ", print=True)
    208 
--> 209     signaling_client = register(url=network_url)
    210 
    211     info("♫♫♫ > " + bcolors.OKGREEN + "DONE!" + bcolors.ENDC, print=True)

/Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/site-packages/syft-0.0.0-py3.9.egg/syft/grid/duet/om_signaling_client.py in register(url)
     13 ) -> SignalingClient:
     14 
---> 15     client = SignalingClient(
     16         url=url, conn_type=HTTPConnection, client_type=NetworkClient
     17     )

/Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/site-packages/syft-0.0.0-py3.9.egg/syft/grid/duet/signaling_client.py in __init__(self, url, conn_type, client_type)
     35             client_id,
     36         ) = client_type.deserialize_client_metadata_from_node(
---> 37             metadata=conn._get_metadata()
     38         )
     39 

/Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/site-packages/syft-0.0.0-py3.9.egg/syft/grid/connections/http_connection.py in _get_metadata(self)
    108         :rtype: str of bytes
    109         """
--> 110         data: bytes = requests.get(self.base_url + "/metadata").content
    111         metadata_pb = Metadata_PB()
    112         metadata_pb.ParseFromString(data)

/Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/site-packages/requests/api.py in get(url, params, **kwargs)
     74 
     75     kwargs.setdefault('allow_redirects', True)
---> 76     return request('get', url, params=params, **kwargs)
     77 
     78 

/Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/site-packages/requests/api.py in request(method, url, **kwargs)
     59     # cases, and look like a memory leak in others.
     60     with sessions.Session() as session:
---> 61         return session.request(method=method, url=url, **kwargs)
     62 
     63 

/Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/site-packages/requests/sessions.py in request(self, method, url, params, data, headers, cookies, files, auth, timeout, allow_redirects, proxies, hooks, stream, verify, cert, json)
    540         }
    541         send_kwargs.update(settings)
--> 542         resp = self.send(prep, **send_kwargs)
    543 
    544         return resp

/Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/site-packages/requests/sessions.py in send(self, request, **kwargs)
    653 
    654         # Send the request
--> 655         r = adapter.send(request, **kwargs)
    656 
    657         # Total elapsed time of the request (approximately)

/Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/site-packages/requests/adapters.py in send(self, request, stream, timeout, verify, cert, proxies)
    514                 raise SSLError(e, request=request)
    515 
--> 516             raise ConnectionError(e, request=request)
    517 
    518         except ClosedPoolError as e:

ConnectionError: HTTPConnectionPool(host='ec2-18-216-8-163.us-east-2.compute.amazonaws.com', port=5000): Max retries exceeded with url: //metadata (Caused by NewConnectionError('<urllib3.connection.HTTPConnection object at 0x7fa7c4a7e460>: Failed to establish a new connection: [Errno 60] Operation timed out'))

In [ ]:

​

**Comment by znreza** [comment]
Hi @raaghavvd Could you solve that issue? It could be possible that the node was down at that moment. Could you please try `"loopback=True"` instead like shown in [this](https://github.com/OpenMined/courses/tree/foundations-of-private-computation/federated-learning/duet_iris_classifier) example? 

Basically change this parameter in both DO and DS notebooks like below

In DO notebook, 
`duet = sy.launch_duet(loopback=True)`

In DS notebook,
`duet = sy.join_duet(loopback=True)
`
Hope it works!

**Reply by owls-wonderland** [reply]
Hi! I have been having the same problem for some time, before i found this answer. Very happy now!

However I have a question. So if we assume it would be "real-life" situation I would still have to use the link, right? How do one get this link? I would appreciate your answer.
Thank you!

**Reply by LeonMac** [reply]
I have a similar issue (RemoteDisconnected ) that looks the duet server is down right now?
I use the solution as proposed -- this works OK. this is likely a loopback mode that DO and DS simply do communication directly so there is no token stuff happening at all.

> In DO notebook,
> duet = sy.launch_duet(loopback=True)
> 
> In DS notebook,
> duet = sy.join_duet(loopback=True)

**Reply by yemikifouly** [reply]
@LeonMac Please, consider asking this question in the #support channel of our [Slack workspace](www.slack.openmined.org).
